engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java

```diff
@@ -40,6 +40,20 @@ public final class ArraySlice implements TruffleObject {
     }
   }
 
+  static Object create(Object storage, long start, long this_length, long end) {
+    long slice_start = Math.max(0, start);
+    long slice_end = Math.min(this_length, end);
+    Object slice;
+    if (slice_start >= slice_end) {
+      slice = Array.allocate(0);
+    } else if ((slice_start == 0) && (slice_end == this_length)) {
+      slice = null;
```

Comment:
This is confusing. I see what is going on, but it's a really 'magical' use of `null` to indicate that we do not need to allocate a new Vector.

Why can't we just do the `Vector.fromArray` in this method? And just return `storage` in this branch then?

If you think passing the magic `null` value is best - OK - but please add a comment to the function. It's very confusing that a function returning a slice of an array returns `null` when the selected slice is the full array - that's definitely not what I'd expect from a `ArraySlice.create` call. So it would be good to add a note about it.

Reply:
I see. I think https://github.com/enso-org/enso/commit/2de4f39212513a03391fefc11492e07d1c282e20 looks good, but I'd add a comment to `createOrNull` which would say what it means that it returns `null`.